tests/lib/prepare.sh

```diff
@@ -365,41 +365,50 @@ EOF
         # build the user database - this is complicated because:
         # - spread on linode wants to login as "root"
         # - "root" login on the stock core snap is disabled
-        # - we need to add our ubuntu and test users too
         # - uids between classic/core differ
         # - passwd,shadow on core are read-only
         # - we cannot add root to extrausers as system passwd is searched first
-        # So we do:
-        # - take core passwd without "root" as extrausers
-        # - append root,ubuntu,test to extrausers
-        # - bind mount extrausers to /etc via custom systemd job
+        # - we need to add our ubuntu and test users too
+        # So we create the user db we need in /root/test-etc/*:
+        # - take core passwd without "root"
+        # - append root
+        # - make sure the group matches
+        # - bind mount /root/test-etc/* to /etc/* via custom systemd job
+        # We also create /var/lib/extrausers/* and append ubuntu,test there
+        mkdir -p /mnt/system-data/root/test-etc
         mkdir -p /mnt/system-data/var/lib/extrausers/
         touch /mnt/system-data/var/lib/extrausers/sub{uid,gid}
         mkdir -p /mnt/system-data/etc/systemd/system/multi-user.target.wants
         for f in group gshadow passwd shadow; do
             # the passwd from core without root
-            tail -n +2 "$UNPACKD/etc/$f" > /mnt/system-data/var/lib/extrausers/$f
+            tail -n +2 "$UNPACKD/etc/$f" > /mnt/system-data/root/test-etc/$f
             # append this systems root user so that linode can connect
-            head -n1 /etc/$f >> /mnt/system-data/var/lib/extrausers/$f
-            # append ubuntu, test user for the testing
-            tail -n2 /etc/$f >> /mnt/system-data/var/lib/extrausers/$f
+            head -n1 /etc/$f >> /mnt/system-data/root/test-etc/$f
 
-            # now bind mount those passwd files on boot
+            # make sure the group is as expected
+            chgrp --reference "$UNPACKD/etc/$f" /mnt/system-data/root/test-etc/$f
+            # now bind mount read-only those passwd files on boot
             cat <<EOF > /mnt/system-data/etc/systemd/system/etc-$f.mount
 [Unit]
-Description=Mount extrausers $f over system $f
+Description=Mount root/test-etc/$f over system etc/$f
 Before=ssh.service
 
 [Mount]
-What=/var/lib/extrausers/$f
+What=/root/test-etc/$f
 Where=/etc/$f
 Type=none
-Options=bind
+Options=bind,ro
 
 [Install]
 WantedBy=multi-user.target
 EOF
             ln -s /etc/systemd/system/etc-$f.mount /mnt/system-data/etc/systemd/system/multi-user.target.wants/etc-$f.mount
+
+            # create /var/lib/extrausers/$f
+            # append ubuntu, test user for the testing
+            cp -a "$UNPACKD/var/lib/extrausers/$f" /mnt/system-data/var/lib/extrausers/$f
+            tail -n2 /etc/$f >> /mnt/system-data/var/lib/extrausers/$f
```

Comment:
I know this is existing code, just slightly re-organized, but I'm slightly concerned about the `tail -n2 ..` and `head -n1 ..` approach used in this script, seems fragile and built around assumption that the right stuff is at the right place and never changes. Can we make it slightly more robust or add a sanity check here in prepare to ensure expected users are present (before failing in misterious ways later?)

Reply:
as we discussed I can probably switch to a mixed usage of grep instead of tail/head or for sanity checking, but in a follow up